src/bin/scribble.ts

```diff
@@ -651,327 +705,314 @@ if ("version" in options) {
+            writeOut(resultJSON, options.output);
+        }
+    } else {
+        if (options["arm"] && isMetaDataFile) {
+            error(
+                `Instrumentation file "${metaDataFile}" already exists. Consider disarming or providing other path for metadata file.`
+            );
+        }
 
-                if (!options.quiet) {
-                    console.error(`${unit.absolutePath} -> ${instrumentedFileName}`);
-                }
+        modifiedFiles = [...instrCtx.changedUnits, utilsUnit];
+        // 1. In 'files' mode first write out the files
+        const newContents = instrCtx.printUnits(modifiedFiles, newSrcMap, instrumentationMarker);
 
-                fse.writeFileSync(instrumentedFileName, newContents.get(unit) as string);
+        // 2. For all changed files write out a `.instrumented` version of the file.
+        for (const unit of instrCtx.changedUnits) {
+            const instrumentedFileName = unit.absolutePath + ".instrumented";
+
+            if (!options.quiet) {
+                console.error(`${unit.absolutePath} -> ${instrumentedFileName}`);
             }
 
-            // 3. Write out the utils contract
-            fse.writeFileSync(utilsUnit.absolutePath, newContents.get(utilsUnit) as string);
+            fse.writeFileSync(instrumentedFileName, newContents.get(unit) as string);
+        }
 
-            // 4. Finally if --arm is passed put the instrumented files in-place
-            if (options["arm"]) {
-                for (const unit of instrCtx.changedUnits) {
-                    const instrumentedFileName = unit.absolutePath + ".instrumented";
-                    const originalFileName = unit.absolutePath + ".original";
+        // 3. Write out the utils contract
+        fse.writeFileSync(utilsUnit.absolutePath, newContents.get(utilsUnit) as string);
 
-                    copy(unit.absolutePath, originalFileName, options);
-                    copy(instrumentedFileName, unit.absolutePath, options);
-                }
+        // 4. Finally if --arm is passed put the instrumented files in-place
+        if (options["arm"]) {
+            for (const unit of instrCtx.changedUnits) {
+                const instrumentedFileName = unit.absolutePath + ".instrumented";
+                const originalFileName = unit.absolutePath + ".original";
+
+                copy(unit.absolutePath, originalFileName, options);
+                copy(instrumentedFileName, unit.absolutePath, options);
             }
         }
+    }
 
-        if (options["instrumentation-metadata-file"] !== undefined) {
-            const metadata: any = generateInstrumentationMetadata(
-                instrCtx,
-                newSrcMap,
-                instrCtx.units,
-                modifiedFiles,
-                options["arm"] !== undefined,
-                pkg.version,
-                options["output"]
-            );
+    if (options["arm"] || options["instrumentation-metadata-file"]) {
```

Comment:
There is a small inconsistency here. In line 586 we throw an error if we can't detect a location for an instrumentation metadata file. We do this no matter the arguments to scribble. However we only really use the metadata files, if we have `--arm`, `--disarm` or `--instrumentation-metadata-file`. I am wondering if we should either:

a) Always emit the metadata file

 - or -
 
b) Only require an `--instrumentation-metadata-file` if we can't detect project root, and there is an `--arm` or `--disarm`

What do you think?

Reply:
Always emitting metadata does not match with the flat / JSON output. All operations are happening in memory and nothing is actually written to any files. Emitting the metadata file in such case causing confusion because Scribble starting to complaing that can not do same thing due to there is a metadata file, that can not be overwritten for different inputs. Currently Scribble follows option B from your list.

Reply:
Thats fair. I agree with you.